File: hsg12-tinh/2024-2025-lamdong/buatiec/buatiec.cpp
#include <iostream>
#include <vector>
#include <map>
#include <stack>
#include <algorithm>

#define input_file "buatiec.inp"
#define output_file "buatiec.out"

using namespace std;

int n;
vector<int> H;

// Danh sách kề
map<int, vector<int>> adj;

// Mảng D dùng để lưu tổng độ vui vẻ
// D[l][0] là tổng độ vui vẻ lớn nhất nếu l không được chọn
// D[l][1] là tổng độ vui vẻ lớn nhất nếu l được chọn
vector<vector<int>> D;


void input()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    freopen(input_file, "r", stdin);

    cin >> n;

    H.resize(n + 1);

    for (int i = 1; i < n + 1; ++i)
    {
        cin >> H[i];
    }

    int l, k;
    for (int i = 1; i < n; ++i)
    {
        cin >> l >> k;
        adj[k].push_back(l);
    }
}


// Hàm dùng để duyệt cây theo chiều sâu
void dfs(int k)
{
    // Khởi tạo khi chọn node k
    D[k][1] = H[k];

    // Duyệt từng con của node k
    for (int l : adj[k])
    {
        // Gọi đệ quy
        dfs(l);

        // Nếu chọn k thì không thể chọn con của k
        D[k][1] += D[l][0];

        // Nếu không chọn k thì có thể chọn hoặc không chọn con của k
        D[k][0] += max(D[l][0], D[l][1]);
    }
}


void process()
{
    // Khởi tạo D gồm n+1 hàng và 2 cột
    D.resize(n + 1, vector<int>(2, 0));

    // Gọi hàm dfs từ node 1 (giám đốc)
    dfs(1);

    // Tổng độ vui vẻ lớn nhất
    int result = D[1][1];

    // In kết quả
    freopen(output_file, "w", stdout);
    cout << result;
}


int main()
{
    input();
    process();

    return 0;
}

File: hsg12-tinh/2024-2025-lamdong/buatiec/buatiec.py
import os
import sys

input_file = os.path.join(sys.path[0], 'buatiec.inp')
output_file = os.path.join(sys.path[0], 'buatiec.out')

n = 0
H = []

# Danh sách kề
adj = dict()

# Mảng D dùng để lưu tổng độ vui vẻ
# D[l][0] là tổng độ vui vẻ lớn nhất nếu l không được chọn
# D[l][1] là tổng độ vui vẻ lớn nhất nếu l được chọn
D = []


def input_data():
    global n, H, adj

    with open(input_file, 'r') as file:
        n = int(file.readline().strip())
        H = list(map(int, file.readline().split()))

        # Thêm 0 vào đầu H để phù hợp với chỉ số của node
        H.insert(0, 0)

        for i in range(n - 1):
            l, k = map(int, file.readline().split())
            if k not in adj:
                adj[k] = []
            adj[k].append(l)


# Hàm dùng để duyệt cây theo chiều sâu
def dfs(k):
    global D, H

    # Khởi tạo khi chọn node k
    D[k][1] = H[k]

    # Duyệt từng con của node k
    for l in adj.get(k, []):
        # Gọi đệ quy
        dfs(l)

        # Nếu chọn k thì không thể chọn con của k
        D[k][1] += D[l][0]

        # Nếu không chọn k thì có thể chọn hoặc không chọn con của k
        D[k][0] += max(D[l][0], D[l][1])


def process():
    global D

    # Khởi tạo D gồm n+1 hàng và 2 cột
    D = [[0, 0] for _ in range(n + 1)]

    # Gọi hàm dfs từ node 1 (giám đốc)
    dfs(1)

    # Tổng độ vui vẻ lớn nhất
    result = D[1][1]

    with open(output_file, 'w') as file:
        file.write(str(result))


if __name__ == '__main__':
    input_data()
    process()